app/controllers/users.rb

```diff
@@ -0,0 +1,36 @@
+# get a list of users
+get '/users' do
+end
+# get a form to add a new users
+get '/users/new' do
+
+  erb :new_user
+end
+# add a new users
+post '/users' do
+  session[:errors] = nil
+  user = User.create(username: params[:username], email: params[:email], password: params[:password])
+  if user.errors.any?
+    session[:errors] = user.display_errors
+    redirect '/users/new' #<< partial (: invalid_user )
+  elsif
+    session[:user_id] = user.id
```

Comment:
Common session modifiations like `session[:user_id] = @user.id` should be moved into a helper like so:

```ruby
helpers do
  def login_user(user)
    session[:user_id] = user.id
  end
end
```